### PR TITLE
[Labs/Select] Fix undesired page scroll when focusing input in popover

### LINKED ISSUE #1582: [Labs/Select] Page scrolls when popover opens
When you click on the button and the popover opens, the page scrolls all the way back up.

### PR DESCRIPTION
**Pull request by brieb**
#### Fixes #1582

To be merged into https://github.com/palantir/blueprint/pull/1579 (bb/1512)

### CODE REVIEW

**Comment by brieb** [comment]
Looked like this was caused by the combination of autofocus and animation.
Couldn't find much on the topic, but found...
https://bugs.chromium.org/p/chromium/issues/detail?id=360488
https://stackoverflow.com/questions/26782998/why-does-calling-focus-break-my-css-transition

I tested this by rebasing on top of https://github.com/palantir/blueprint/pull/1579 and making sure that the page didn't scroll when the popover2 opened.

File: packages/labs/src/components/select/select.tsx [hunk 1]
```diff
@@ -114,9 +114,18 @@ export class Select<T> extends React.Component<ISelectProps<T>, ISelectState<T>>
...
 
+    private input: HTMLInputElement;
     private TypedQueryList = QueryList.ofType<T>();
     private list: QueryList<T>;
     private refHandlers = {
+        input: (ref: HTMLInputElement) => {
+            this.input = ref;
+
+            const { inputProps = {} } = this.props;
+            if (inputProps.inputRef !== undefined) {
```
Comment (cmslewis) [inline packages/labs/src/components/select/select.tsx]: Can use `safeInvoke`. Looks like we already import core `Utils` into this file, but this is how I prefer to do it in `packages/table` at least, given the overloaded `utils.ts` filename that appears in both `core` and `table`:

```tsx
import { Utils as CoreUtils } from "@blueprintjs/core";
// ...
const { inputProps = {} } = this.props;
CoreUtils.safeInvoke(inutProps.inputRef, ref);
```

File: packages/labs/src/components/select/select.tsx [hunk 2]
```diff
@@ -269,6 +279,12 @@ export class Select<T> extends React.Component<ISelectProps<T>, ISelectState<T>>
             this.list.scrollActiveItemIntoView();
         }
 
+        requestAnimationFrame(() => {
```
Comment (cmslewis) [inline packages/labs/src/components/select/select.tsx]: _Snazzy_. 👏 Though this seems brittle. Can you write a unit test for this to guarantee the behavior? Probably something like:

```tsx
it("should focus input when popover opened if autofocus=true", (done) => {
    // setup, open popover
    // use setTimeout to wait for the focus to transfer within requestAnimationFrame
    setTimeout(() => {
        // check: is popover input focused?
        done();
    });
}); 
```

**Comment by llorca** [comment]
This is a good approach, but just to clarify, is there an actual regression between `Popover` and `Popover2` that needs to be addressed? Why was it not a problem with `Popover`?

@giladgray would love your thoughts as well since you built `Popover2`

File: packages/labs/src/components/select/select.tsx (hunk 2 above)
Comment (giladgray) [reply packages/labs/src/components/select/select.tsx]: this should happen in the ref handler itself, not on every render.

`autoFocus` only applies when the component appears, not each time it updates.

> This Boolean attribute lets you specify that a form control should have input focus **when the page loads** ([MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input))

File: packages/labs/src/components/select/select.tsx [hunk 3]
```diff
@@ -159,12 +168,13 @@ export class Select<T> extends React.Component<ISelectProps<T>, ISelectState<T>>
         const { ref, ...htmlInputProps } = inputProps;
         const input = (
             <InputGroup
-                autoFocus={true}
+                autoFocus={false}
```
Comment (giladgray) [inline packages/labs/src/components/select/select.tsx]: remove this line. this is the default and user can override with spread anyway.

File: packages/labs/src/components/select/select.tsx (hunk 2 above)
Comment (giladgray) [reply packages/labs/src/components/select/select.tsx]: might be worth respecting `inputProps.autoFocus === false`? I could go either way.

Comment (brieb) [reply packages/labs/src/components/select/select.tsx]: Is there a good way to test for this when we use requestAnimationFrame? Was able to write a test that passes in the browser, but not in the cli. Found this comment, which seems related: https://github.com/palantir/blueprint/blob/874b757e729576a330ee3a8b67a382c7ff4a312e/packages/core/test/common/utilsTests.ts#L106-L107

**Comment by brieb** [comment]
re respecting `inputProps.autoFocus === false`: don't see why it is useful, but is technically correct. so, i'll add it

**Comment by llorca** [comment]
Do we expect the prod move from `Popover` to `Popover2` to behave this way with consumer apps too? That is to say, will consumers need to update their code the same way we did here?

File: packages/labs/src/components/select/select.tsx (hunk 2 above)
Comment (giladgray) [reply packages/labs/src/components/select/select.tsx]: just use `setTimeout` to test it on the next frame. 

that `utilsTest` is for a util that must use `rAF`, but for other focus stuff a timeout is enough.

**Pull request merged**